File: server-python/account_server/account_server.py
"""账号服 HTTP 服务（:9000）：注册 / 登录 / 游客 / 微信登录 / 头像代理。

对应 `server/account_server/account_server.ts`。

与 Node 版一致，本文件**没有**跟进 `utils/http.py` 的统一出口，
而是保留自己的本地 `send_json()`（返回结构也比大厅服/游戏服随意）。

两处历史 bug 忠实保留（移植不修，要修请单独开一次改动）：

1. `/register` 的判断是反的——`is_user_exist` 为真（账号**已存在**）才去创建；
2. `/auth` 调用了一个本文件从未定义的 `get_md5`，那一分支必然抛 NameError
   （Node 版是 ReferenceError）。
"""

from __future__ import annotations

import json
from typing import Any

from aiohttp import web

from utils import crypto, db, http

APP_WEB_DEFAULT = "http://fir.im/2f17"


def send_json(payload: dict[str, Any]) -> web.Response:
    """本地 JSON 出口：账号服的历史约定（不要改成 utils/http 的 `send`）。"""
    return http.json_response(payload)


def _get_md5(content: str) -> str:
    """历史 bug 的忠实保留（见 `/auth` 处理器的说明）。

    原 `account_server.js` 的 `/auth` 分支调用了一个**本文件从未定义**的 `get_md5`，
    因此那一分支必然抛 ReferenceError。移植的目标是行为不变，所以这里不是"补上实现"，
    而是把这个必然抛错的事实写成一个能通过检查的桩：调用点、抛错类型与原实现一致
    （Python 里对应 NameError）。
    """
    raise NameError("name 'get_md5' is not defined")


#: 微信开放平台的应用信息：键是 `/wechat_auth` 的 os 参数（Android / iOS）。
_APP_INFO: dict[str, dict[str, str] | None] = {
    "Android": {
        "appid": "wxe39f08522d35c80d",
        "secret": "fa88e3a3ca5a11b06499902cea4b9c01",
    },
    "iOS": {
        "appid": "wxcb508816c5c4e2a4",
        "secret": "7de38489ede63089269e3410d5905038",
    },
}


def create_routes(config: dict[str, Any], hall_addr: str) -> web.Application:
    """建账号服的 aiohttp 应用。

    :param config: 账号服配置（`configs_*.py` 的 `account_server()`）。
    :param hall_addr: `HALL_IP:HALL_CLIENT_PORT`，`/guest` 与 `/get_serverinfo` 会下发它。
    """
    app = web.Application()

    @web.middleware
    async def cors(request: web.Request, handler: Any) -> web.StreamResponse:
        # 设置跨域访问（与原实现的中间件逐项对应）
        response = await handler(request)
        response.headers["Access-Control-Allow-Origin"] = "*"
        response.headers["Access-Control-Allow-Headers"] = "X-Requested-With"
        response.headers["Access-Control-Allow-Methods"] = "PUT,POST,GET,DELETE,OPTIONS"
        response.headers["X-Powered-By"] = " 3.2.1"
        if "Content-Type" not in response.headers:
            response.headers["Content-Type"] = "application/json;charset=utf-8"
        return response

    app.middlewares.append(cors)

    async def on_register(request: web.Request) -> web.Response:
        account = http.query_string(request, "account")
        password = http.query_string(request, "password")

        def fn_failed() -> web.Response:
            return send_json({"errcode": 1, "errmsg": "account has been used."})

        def fn_succeed() -> web.Response:
            return send_json({"errcode": 0, "errmsg": "ok"})

        # 历史 bug 忠实保留：这里的判断是反的（存在才创建）。
        if await db.is_user_exist(account):
            if await db.create_account(account, password):
                return fn_succeed()
            return fn_failed()
        print("account has been used.")
        return fn_failed()

    async def on_get_version(request: web.Request) -> web.Response:
        return send_json({"version": config["VERSION"]})

    async def on_get_serverinfo(request: web.Request) -> web.Response:
        return send_json(
            {
                "version": config["VERSION"],
                "hall": hall_addr,
                "appweb": config["APP_WEB"],
            }
        )

    async def on_guest(request: web.Request) -> web.Response:
        account = "guest_" + str(http.query_string(request, "account"))
        sign = crypto.md5(account + http.request_ip(request) + config["ACCOUNT_PRI_KEY"])
        return send_json(
            {
                "errcode": 0,
                "errmsg": "ok",
                "account": account,
                "halladdr": hall_addr,
                "sign": sign,
            }
        )

    async def on_auth(request: web.Request) -> web.Response:
        account_in = http.query_string(request, "account")
        password = http.query_string(request, "password")

        info = await db.get_account_info(account_in, password)
        if info is None:
            return send_json({"errcode": 1, "errmsg": "invalid account"})

        account = "vivi_" + str(http.query_string(request, "account"))
        # 历史 bug（移植不修）：原实现调用的是本文件从未定义的 get_md5()。
        # 这行位于 db 的结果之后，抛出的 NameError 会冒泡到 aiohttp（Node 版里
        # 它不在 express 的异常链上，会直接结束账号服进程）。
        # 迁移只把"必然抛错"这个事实写成可读的桩函数，不改行为；要修请单独开一次改动
        # （改用 crypto.md5）。
        sign = _get_md5(account + http.request_ip(request) + config["ACCOUNT_PRI_KEY"])
        return send_json({"errcode": 0, "errmsg": "ok", "account": account, "sign": sign})

    async def get_access_token(code: str, os_name: str) -> http.HttpResult:
        info = _APP_INFO.get(os_name)
        if info is None:
            # 原实现这里**没有 return**：它先回调失败，然后继续往下走并在 info.appid 处
            # 抛异常。净效果是"未知 os 时抛 TypeError"，这里保留同样的取值路径。
            print("unknown os: " + os_name)
        data = {
            "appid": info["appid"],  # type: ignore[index]
            "secret": info["secret"],  # type: ignore[index]
            "code": code,
            "grant_type": "authorization_code",
        }
        return await http.get2("https://api.weixin.qq.com/sns/oauth2/access_token", data, True)

    async def get_state_info(access_token: str, openid: str) -> http.HttpResult:
        data = {"access_token": access_token, "openid": openid}
        return await http.get2("https://api.weixin.qq.com/sns/userinfo", data, True)

    async def create_user(account: str, name: str, sex: int, headimgurl: str) -> None:
        coins = 1000
        gems = 21
        if not await db.is_user_exist(account):
            await db.create_user(account, name, coins, gems, sex, headimgurl)
        else:
            await db.update_user_info(account, name, headimgurl, sex)

    async def on_wechat_auth(request: web.Request) -> web.Response:
        code = http.query_string(request, "code")
        os_name = http.query_string(request, "os")
        if code is None or code == "" or os_name is None or os_name == "":
            return web.Response(text="")
        print(os_name)

        result = await get_access_token(code, os_name)
        if result.ok:
            access_token = result.data.get("access_token")
            openid = result.data.get("openid")
            result2 = await get_state_info(str(access_token), str(openid))
            if result2.ok:
                openid = result2.data.get("openid")
                nickname = result2.data.get("nickname")
                sex = result2.data.get("sex")
                headimgurl = result2.data.get("headimgurl")
                account = "wx_" + str(openid)
                await create_user(account, str(nickname), int(sex), str(headimgurl))
                sign = crypto.md5(account + http.request_ip(request) + config["ACCOUNT_PRI_KEY"])
                return send_json(
                    {
                        "errcode": 0,
                        "errmsg": "ok",
                        "account": account,
                        "halladdr": hall_addr,
                        "sign": sign,
                    }
                )
            return web.Response(text="")
        return send_json({"errcode": -1, "errmsg": "unkown err."})

    async def on_base_info(request: web.Request) -> web.Response:
        userid = http.query_string(request, "userid")
        data = await db.get_user_base_info(userid)
        # 原实现不做空判断：userid 缺失时 db 会回调 null，这里同样在取字段时抛
        # TypeError（行为不变）。
        return send_json(
            {
                "errcode": 0,
                "errmsg": "ok",
                "name": data["name"],  # type: ignore[index]
                "sex": data["sex"],  # type: ignore[index]
                "headimgurl": data["headimg"],  # type: ignore[index]
            }
        )

    async def on_image(request: web.Request) -> web.Response:
        url = http.query_string(request, "url")
        if not url:
            return http.send(1, "invalid url", {})
        if not url.startswith("http://") and not url.startswith("https://"):
            return http.send(1, "invalid url", {})

        url = url.split(".jpg")[0]

        # 原实现据此决定用 http 还是 https；aiohttp 直接从 URL 的 scheme 选协议，
        # 这里保留这行是为了让取值路径与 Node 版读起来一致。
        _safe = url.startswith("https://")
        print(url)
        # 代理拉取远程图片：拿到原始字节后再回写响应。
        content_type, data = await http.get_bytes(url)
        if not content_type or not data:
            # 原实现传的是 `true`：迁移前的 http.js 是非严格模式，给 boolean 挂
            # errcode/errmsg 只是静默失效，响应体因此是 JSON.stringify(true) == "true"。
            # 这里直接写出原来的响应体：对外行为逐字不变。
            return web.Response(text=json.dumps(True), content_type="application/json", charset="utf-8")
        return web.Response(body=data, content_type=content_type)

    app.router.add_get("/register", on_register)
    app.router.add_get("/get_version", on_get_version)
    app.router.add_get("/get_serverinfo", on_get_serverinfo)
    app.router.add_get("/guest", on_guest)
    app.router.add_get("/auth", on_auth)
    app.router.add_get("/wechat_auth", on_wechat_auth)
    app.router.add_get("/base_info", on_base_info)
    app.router.add_get("/image", on_image)
    return app

File: server/account_server/account_server.ts
import * as crypto from "../utils/crypto";
import express from "express";
import type { Request, Response, NextFunction } from "express";
import type { Server } from "node:http";
import * as db from "../utils/db";
import * as http from "../utils/http";
import type { AccountServerConfig } from "../types/config";
import type { AccountRow, UserBaseInfoRow } from "../types/db_rows";

/**
 * 账号服 HTTP 服务（:9000）：注册 / 登录 / 游客 / 微信登录 / 头像代理。
 *
 * 迁移说明（见 docs/ai-native/typescript-migration.md §4、server/AGENTS.md §7.5）：
 * 账号服历史上没有跟进 `utils/http.ts` 的统一出口，本文件保留自己的本地 `send(res, ret)`，
 * 返回结构也比大厅服/游戏服随意；这里只补类型，不改返回结构。
 */

var app = express();
var hallAddr = "";

// 本地 JSON 出口：账号服的历史约定（不要改成 utils/http.ts 的 send）
function send(res: Response,ret: Record<string, unknown>): void {
	var str = JSON.stringify(ret);
	res.send(str)
}

/**
 * 历史 bug 的忠实保留（见 `/auth` 处理器的说明）。
 *
 * 原 `account_server.js` 的 `/auth` 分支调用了一个**本文件从未定义**的 `get_md5`，
 * 因此那一分支必然抛 ReferenceError。迁移的目标是行为不变，所以这里不是"补上实现"，
 * 而是把这个必然抛错的事实写成一个能通过编译的桩：调用点、抛错类型与原实现一致。
 *
 * @param content 原实现想签名的内容（本桩不使用）。
 * @returns 永不返回。
 */
function get_md5(content: string): string {
	throw new ReferenceError("get_md5 is not defined");
}

// 由 start() 在 listen 之前赋值；原实现写作 var config = null
var config: AccountServerConfig;

/**
 * 启动账号服 HTTP 服务。
 *
 * @param cfg 账号服配置（configs_*.ts 的 account_server()）。
 * @returns http.Server，交给 account_server/app.js 汇总成启动横幅。
 */
export function start(cfg: AccountServerConfig): Server {
	config = cfg;
	hallAddr = config.HALL_IP  + ":" + config.HALL_CLIENT_PORT;
	// 返回 http.Server，交给 app.js 汇总成启动横幅（见 utils/startup.js）
	return app.listen(config.CLIENT_PORT);
}




//设置跨域访问
app.all('*', function(req: Request, res: Response, next: NextFunction) {
    res.header("Access-Control-Allow-Origin", "*");
    res.header("Access-Control-Allow-Headers", "X-Requested-With");
    res.header("Access-Control-Allow-Methods","PUT,POST,GET,DELETE,OPTIONS");
    res.header("X-Powered-By",' 3.2.1')
	res.header("Content-Type", "application/json;charset=utf-8");
	next();
});

app.get('/register',function(req: Request,res: Response){
	var account = http.queryString(req, "account");
	var password = http.queryString(req, "password");

	var fnFailed = function(): void {
		send(res,{errcode:1,errmsg:"account has been used."});
	};

	var fnSucceed = function(): void {
		send(res,{errcode:0,errmsg:"ok"});	
	};

	db.is_user_exist(account,function(exist: boolean){
		if(exist){
			db.create_account(account,password,function(ret: boolean){
				if (ret) {
					fnSucceed();
				}
				else{
					fnFailed();
				}
			});
		}
		else{
			fnFailed();
			console.log("account has been used.");			
		}
	});
});

app.get('/get_version',function(req: Request,res: Response){
	var ret = {
		version:config.VERSION,
	}
	send(res,ret);
});

app.get('/get_serverinfo',function(req: Request,res: Response){
	var ret = {
		version:config.VERSION,
		hall:hallAddr,
		appweb:config.APP_WEB,
	}
	send(res,ret);
});

app.get('/guest',function(req: Request,res: Response){
	var account = "guest_" + http.queryString(req, "account");
	// 原实现直接拼 req.ip：express 的类型把它标成 string | undefined，
	// 这里与迁移前一样按字符串用（断言不产生任何运行时代码），不加存在性判断，行为逐字不变
	var sign = crypto.md5(account + (req.ip as string) + config.ACCOUNT_PRI_KEY);
	var ret = {
		errcode:0,
		errmsg:"ok",
		account:account,
		halladdr:hallAddr,
		sign:sign
	}
	send(res,ret);
});

app.get('/auth',function(req: Request,res: Response){
	var account = http.queryString(req, "account");
	var password = http.queryString(req, "password");

	db.get_account_info(account,password,function(info: AccountRow | null){
		if(info == null){
			send(res,{errcode:1,errmsg:"invalid account"});
			return;
		}

        var account = "vivi_" + http.queryString(req, "account");
        // 历史 bug（迁移不修）：原实现调用的是本文件从未定义的 get_md5()。
        // 这行位于 db 的异步回调里，抛出的 ReferenceError 不会被 express 捕获，账号服进程会直接退出。
        // 迁移只把"必然抛错"这个事实写成可读的桩函数，不改行为；要修请单独开一次改动（改用 crypto.md5）。
        var sign = get_md5(account + (req.ip as string) + config.ACCOUNT_PRI_KEY);
        var ret = {
            errcode:0,
            errmsg:"ok",
            account:account,
            sign:sign
        }
        send(res,ret);
	});
});

/** 微信开放平台的应用信息：键是 /wechat_auth 的 os 参数（Android / iOS）。 */
interface WechatApp {
	appid: string;
	secret: string;
}

var appInfo: Record<string, WechatApp | undefined> = {
	Android:{
		appid:"wxe39f08522d35c80c",
		secret:"fa88e3a3ca5a11b06499902cea4b9c01",
	},
	iOS:{
		appid:"wxcb508816c5c4e2a4",
		secret:"7de38489ede63089269e3410d5905038",		
	}
};

function get_access_token(code: string,os: string,callback: http.HttpCallback): void {
	var info = appInfo[os];
	if(info == null){
		// 原实现这里没有 return，回调之后还会继续往下走并在 info.appid 处抛异常；
		// info! 保留同样的取值路径（未知 os 时同样抛 TypeError），只用于通过类型检查
		callback({ ok: false, error: new Error("unknown os: " + os) });
	}
	var data: http.QueryParams = {
		appid:info!.appid,
		secret:info!.secret,
		code:code,
		grant_type:"authorization_code"
	};

	http.get2("https://api.weixin.qq.com/sns/oauth2/access_token",data,callback,true);
}

function get_state_info(access_token: string,openid: string,callback: http.HttpCallback): void {
	var data: http.QueryParams = {
		access_token:access_token,
		openid:openid
	};

	http.get2("https://api.weixin.qq.com/sns/userinfo",data,callback,true);
}

function create_user(account: string,name: string,sex: number,headimgurl: string,callback: () => void): void {
	var coins = 1000;
	var gems = 21;
	db.is_user_exist(account,function(ret: boolean){
		if(!ret){
			db.create_user(account,name,coins,gems,sex,headimgurl,function(ret: unknown){
				callback();
			});
		}
		else{
			db.update_user_info(account,name,headimgurl,sex,function(ret: unknown){
				callback();
			});
		}
	});
};
app.get('/wechat_auth',function(req: Request,res: Response){
	var code = http.queryString(req, "code");
	var os = http.queryString(req, "os");
	if(code == null || code == "" || os == null || os == ""){
		return;
	}
	console.log(os);
	get_access_token(code,os,function(result: http.HttpResult){
		if(result.ok){
			// 微信接口的字段在 JsonObject 里是 unknown，按接口契约在边界处收窄
			var access_token = result.data.access_token as string;
			var openid = result.data.openid as string;
			get_state_info(access_token,openid,function(result2: http.HttpResult){
				if(result2.ok){
					var openid = result2.data.openid as string;
					var nickname = result2.data.nickname as string;
					var sex = result2.data.sex as number;
					var headimgurl = result2.data.headimgurl as string;
					var account = "wx_" + openid;
					create_user(account,nickname,sex,headimgurl,function(){
						var sign = crypto.md5(account + (req.ip as string) + config.ACCOUNT_PRI_KEY);
					    var ret = {
					        errcode:0,
					        errmsg:"ok",
					        account:account,
					        halladdr:hallAddr,
					        sign:sign
					    };
					    send(res,ret);
					});						
				}
			});
		}
		else{
			send(res,{errcode:-1,errmsg:"unkown err."});
		}
	});
});

app.get('/base_info',function(req: Request,res: Response){
	var userid = http.queryString(req, "userid");
	db.get_user_base_info(userid,function(data: UserBaseInfoRow | null){
		// 原实现不做空判断：userid 缺失时 db 会回调 null，这里同样在取字段时抛 TypeError（行为不变），
		// data! 只用于通过类型检查
		var ret = {
	        errcode:0,
	        errmsg:"ok",
			name:data!.name,
			sex:data!.sex,
	        headimgurl:data!.headimg
	    };
	    send(res,ret);
	});
});

app.get('/image', function (req: Request, res: Response) {
	var url = http.queryString(req, "url");
	if (!url) {
	  http.send(res, 1, 'invalid url', {});
	  return;
	}
	if(url.search('http://') != 0 && url.search('https://') != 0){
		http.send(res, 1, 'invalid url', {});
		return;
	}

	url = url.split('.jpg')[0];
	

	var safe = url.search('https://') == 0;
	console.log(url);
	// 代理拉取远程图片：异步回调写法，拿到原始字节后再回写响应。
	// 原实现给 data 传 null：getRaw 的形参类型是 QueryParams，而 querystring.stringify(null)
	// 与 querystring.stringify({}) 都是空串，请求 URL 完全一致。
	http.getRaw(url, {}, safe, 'binary', function (type: string | undefined | null, data: http.JsonObject | string | null) {
		if (!type || !data) {
			// 原实现传的是 `true`：迁移前的 http.js 是非严格模式，给 boolean 挂 errcode/errmsg 只是静默失效，
			// 响应体因此是 JSON.stringify(true) == "true"。迁移后的 http.ts 带 "use strict"，
			// 再走 http.send(res, 1, 'invalid url', true) 会在 getRaw 的异步回调里抛 TypeError
			// （express 捕获不到，会让账号服进程退出）。这里直接写出原来的响应体：
			// 对外行为逐字不变，又不依赖 sloppy mode。
			res.send(JSON.stringify(true));
			return;
		}
		res.writeHead(200, { "Content-Type": type });
		// encoding 是 'binary'，getRaw 在这个分支只会回传字符串；断言只用于收窄 JsonObject | string
		res.write(data as string, 'binary');
		res.end();
	});
});
